finality-aleph/src/data_io/data_interpreter.rs

```diff
@@ -0,0 +1,132 @@
+                // Available(BranchOk) and now the same proposal would be Available(HopelessFork). This is still fine as
+                // the code below will handle such proposals correctly (due to the fact that the branch in such a proposal
+                // is perfectly honest and thus it will be detected as conflicting with previously finalized blocks).
+
+                // WARNING: If we ever enable pruning, this code (and the code in Data Store) must be carefully analyzed
+                // for possible safety violations.
+
+                use crate::data_io::proposal::{AvailableKind::*, ProposalStatus::*};
+                let status = self
+                    .proposal_status_provider
+                    .lock()
+                    .get_status(&proposal, None);
+                match status {
+                    Available(BranchOk) => {
+                        if proposal.number_top_block() <= self.last_finalized_by_aleph.num {
+                            None
+                        } else if let Some(block) =
+                            proposal.block_at_num(self.last_finalized_by_aleph.num)
+                        {
+                            // We must check if the last_finalized_by_aleph block agrees with the one in the proposed branch
+                            // Otherwise this is a hopeless fork that we must ignore.
+                            if block.hash == self.last_finalized_by_aleph.hash {
+                                Some(proposal.top_block())
+                            } else {
+                                None
+                            }
+                        } else {
+                            // The last_finalized_by_aleph block is strictly below the branch bottom block. We apply the proposal,
+                            // as it agrees with last_finalized_by_aleph.
+                            Some(proposal.top_block())
+                        }
+                    }
+                    Available(HopelessFork) => {
+                        debug!(target:"afa", "HopelessFork {:?} encountered in Data. Skipping.", proposal);
+                        None
+                    }
+                    _ => {
+                        // Here we might consider panicking.
+                        error!(target:"afa", "Unavailable proposal {:?} with status {:?} encountered in Data. Skipping.", proposal, status);
+                        None
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl<B: BlockT, PSP: ProposalStatusProvider<B> + Send + 'static>
+    aleph_bft::FinalizationHandler<AlephData<B>> for OrderedDataInterpreter<B, PSP>
+{
+    async fn data_finalized(&mut self, data: AlephData<B>) {
+        if let Some(block) = self.block_to_finalize_from_data(data) {
+            self.last_finalized_by_aleph = block.clone();
```

Comment:
I would be careful with this assignment, as at this point we do not have guarantee that both sending block succeeded and that the finalization happened in the end

Reply:
What we are tracking here (in `last_finalized_by_aleph`) and the finalization in the client are two different things. 

Finalized by aleph means that we must initiliaze the multisigning protocol on this block (aggregator does that). After the multisignature is ready, justification handler gets it and finalizes the block (so essentially if the client says we have a finalized block, it is equivalent to us having a justification for it). 

There is also a very long comment I wrote for @timorl regarding this difference...

I hope that this is what you are confused about here and not something completely unrelated :P